README
NAME

    POE::Component::SSLify - Makes using SSL in the world of POE easy!

VERSION

      This document describes v1.011 of POE::Component::SSLify - released November 13, 2014 as part of POE-Component-SSLify.

SYNOPSIS

            # look at the DESCRIPTION for client and server example code

DESCRIPTION

    This component is a method to simplify the SSLification of a socket
    before it is passed to a POE::Wheel::ReadWrite wheel in your
    application.

 Client usage

            # Import the module
            use POE::Component::SSLify qw( Client_SSLify );
    
            # Create a normal SocketFactory wheel and connect to a SSL-enabled server
            my $factory = POE::Wheel::SocketFactory->new;
    
            # Time passes, SocketFactory gives you a socket when it connects in SuccessEvent
            # Convert the socket into a SSL socket POE can communicate with
            my $socket = shift;
            eval { $socket = Client_SSLify( $socket ) };
            if ( $@ ) {
                    # Unable to SSLify it...
            }
    
            # Now, hand it off to ReadWrite
            my $rw = POE::Wheel::ReadWrite->new(
                    Handle  =>      $socket,
                    # other options as usual
            );

 Server usage

            # !!! Make sure you have a public key + certificate
            # excellent howto: http://www.akadia.com/services/ssh_test_certificate.html
    
            # Import the module
            use POE::Component::SSLify qw( Server_SSLify SSLify_Options );
    
            # Set the key + certificate file
            eval { SSLify_Options( 'server.key', 'server.crt' ) };
            if ( $@ ) {
                    # Unable to load key or certificate file...
            }
    
            # Create a normal SocketFactory wheel to listen for connections
            my $factory = POE::Wheel::SocketFactory->new;
    
            # Time passes, SocketFactory gives you a socket when it gets a connection in SuccessEvent
            # Convert the socket into a SSL socket POE can communicate with
            my $socket = shift;
            eval { $socket = Server_SSLify( $socket ) };
            if ( $@ ) {
                    # Unable to SSLify it...
            }
    
            # Now, hand it off to ReadWrite
            my $rw = POE::Wheel::ReadWrite->new(
                    Handle  =>      $socket,
                    # other options as usual
            );

FUNCTIONS

 Client_SSLify

    This function sslifies a client-side socket. You can pass several
    options to it:

            my $socket = shift;
            $socket = Client_SSLify( $socket, $version, $options, $ctx, $callback );
                    $socket is the non-ssl socket you got from somewhere ( required )
                    $version is the SSL version you want to use
                    $options is the SSL options you want to use
                    $ctx is the custom SSL context you want to use
                    $callback is the callback hook on success/failure of sslification
    
                    # This is an example of the callback and you should pass it as Client_SSLify( $socket, ... , \&callback );
                    sub callback {
                            my( $socket, $status, $errval ) = @_;
                            # $socket is the original sslified socket in case you need to play with it
                            # $status is either 1 or 0; with 1 signifying success and 0 failure
                            # $errval will be defined if $status == 0; it's the numeric SSL error code
                            # check http://www.openssl.org/docs/ssl/SSL_get_error.html for the possible error values ( and import them from Net::SSLeay! )
    
                            # The return value from the callback is discarded
                    }

    If $ctx is defined, SSLify will ignore $version and $options.
    Otherwise, it will be created from the $version and $options
    parameters. If all of them are undefined, it will follow the defaults
    in "SSLify_ContextCreate".

    BEWARE: If you passed in a CTX, SSLify will do Net::SSLeay::CTX_free(
    $ctx ) when the socket is destroyed. This means you cannot reuse
    contexts!

    NOTE: The way to have a client socket with proper certificates set up
    is:

            my $socket = shift;     # get the socket from somewhere
            my $ctx = SSLify_ContextCreate( 'server.key', 'server.crt' );
            $socket = Client_SSLify( $socket, undef, undef, $ctx );

    NOTE: You can pass the callback anywhere in the arguments, we'll figure
    it out for you! If you want to call a POE event, please look into the
    postback/callback stuff in POE::Session.

            # we got this from POE::Wheel::SocketFactory
            sub event_SuccessEvent {
                    my $socket = $_[ARG0];
                    $socket = Client_SSLify( $socket, $_[SESSION]->callback( 'sslify_result' ) );
                    $_[HEAP]->{client} = POE::Wheel::ReadWrite->new(
                            Handle => $socket,
                            ...
                    );
                    return;
            }
    
            # the callback event
            sub event_sslify_result {
                    my ($creation_args, $called_args) = @_[ARG0, ARG1];
                    my( $socket, $status, $errval ) = @$called_args;
    
                    if ( $status ) {
                            print "Yay, SSLification worked!";
                    } else {
                            print "Aw, SSLification failed with error $errval";
                    }
            }

 Server_SSLify

    This function sslifies a server-side socket. You can pass several
    options to it:

            my $socket = shift;
            $socket = Server_SSLify( $socket, $ctx, $callback );
                    $socket is the non-ssl socket you got from somewhere ( required )
                    $ctx is the custom SSL context you want to use; overrides the global ctx set in SSLify_Options
                    $callback is the callback hook on success/failure of sslification

    BEWARE: "SSLify_Options" must be called first if you aren't passing a
    $ctx. If you want to set some options per-connection, do this:

            my $socket = shift;     # get the socket from somewhere
            my $ctx = SSLify_ContextCreate();
            # set various options on $ctx as desired
            $socket = Server_SSLify( $socket, $ctx );

    NOTE: You can use "SSLify_GetCTX" to modify the global, and avoid doing
    this on every connection if the options are the same...

    Please look at "Client_SSLify" for more details on the callback hook.

 SSLify_ContextCreate

    Accepts some options, and returns a brand-new Net::SSLeay context
    object ( $ctx )

            my $ctx = SSLify_ContextCreate( $key, $cert, $version, $options );
                    $key is the certificate key file
                    $cert is the certificate file
                    $version is the SSL version to use
                    $options is the SSL options to use

    You can then call various Net::SSLeay methods on the context

            my $mode = Net::SSLeay::CTX_get_mode( $ctx );

    By default we don't use the SSL key + certificate files

    By default we use the version: default. Known versions of the SSL
    connection - look at http://www.openssl.org/docs/ssl/SSL_CTX_new.html
    for more info.

            * sslv2
            * sslv3
            * tlsv1
            * sslv23
            * default ( sslv23 )

    By default we don't set any options - look at
    http://www.openssl.org/docs/ssl/SSL_CTX_set_options.html for more info.

 SSLify_Options

    Call this function to initialize the global server-side context object.
    This will be the default context whenever you call "Server_SSLify"
    without passing a custom context to it.

            SSLify_Options( $key, $cert, $version, $options );
                    $key is the certificate key file ( required )
                    $cert is the certificate file ( required )
                    $version is the SSL version to use
                    $options is the SSL options to use

    By default we use the version: default

    By default we use the options: Net::SSLeay::OP_ALL

    Please look at "SSLify_ContextCreate" for more info on the available
    versions/options.

 SSLify_GetCTX

    Returns the actual Net::SSLeay context object in case you wanted to
    play with it :)

    If passed in a socket, it will return that socket's $ctx instead of the
    global.

            my $ctx = SSLify_GetCTX();                      # get the one set via SSLify_Options
            my $ctx = SSLify_GetCTX( $sslified_sock );      # get the one in the object

 SSLify_GetCipher

    Returns the cipher used by the SSLified socket

            print "SSL Cipher is: " . SSLify_GetCipher( $sslified_sock ) . "\n";

    NOTE: Doing this immediately after Client_SSLify or Server_SSLify will
    result in "(NONE)" because the SSL handshake is not done yet. The
    socket is nonblocking, so you will have to wait a little bit for it to
    get ready.

            apoc@blackhole:~/mygit/perl-poe-sslify/examples$ perl serverclient.pl
            got connection from: 127.0.0.1 - commencing Server_SSLify()
            SSLified: 127.0.0.1 cipher type: ((NONE))
            Connected to server, commencing Client_SSLify()
            SSLified the connection to the server
            Connected to SSL server
            Input: hola
            got input from: 127.0.0.1 cipher type: (AES256-SHA) input: 'hola'
            Got Reply: hola
            Input: ^C
            stopped at serverclient.pl line 126.

 SSLify_GetSocket

    Returns the actual socket used by the SSLified socket, useful for stuff
    like getpeername()/getsockname()

            print "Remote IP is: " . inet_ntoa( ( unpack_sockaddr_in( getpeername( SSLify_GetSocket( $sslified_sock ) ) ) )[1] ) . "\n";

 SSLify_GetSSL

    Returns the actual Net::SSLeay object so you can call methods on it

            print Net::SSLeay::dump_peer_certificate( SSLify_GetSSL( $sslified_sock ) );

 SSLify_GetStatus

    Returns the status of the SSL negotiation/handshake/connection. See
    http://www.openssl.org/docs/ssl/SSL_connect.html#RETURN_VALUES for more
    info.

            my $status = SSLify_GetStatus( $socket );
                    -1 = still in negotiation stage ( or error )
                     0 = internal SSL error, connection will be dead
                     1 = negotiation successful

NOTES

 Socket methods doesn't work

    The new socket this module gives you actually is tied socket magic, so
    you cannot do stuff like getpeername() or getsockname(). The only way
    to do it is to use "SSLify_GetSocket" and then operate on the socket it
    returns.

 Dying everywhere...

    This module will die() if Net::SSLeay could not be loaded or it is not
    the version we want. So, it is recommended that you check for errors
    and not use SSL, like so:

            eval { use POE::Component::SSLify };
            if ( $@ ) {
                    $sslavailable = 0;
            } else {
                    $sslavailable = 1;
            }
    
            # Make socket SSL!
            if ( $sslavailable ) {
                    eval { $socket = POE::Component::SSLify::Client_SSLify( $socket ) };
                    if ( $@ ) {
                            # Unable to SSLify the socket...
                    }
            }

  $IGNORE_SSL_ERRORS

    As of SSLify v1.003 you can override this variable to temporarily
    ignore some SSL errors. This is useful if you are doing crazy things
    with the underlying Net::SSLeay stuff and don't want to die. However,
    it won't ignore all errors as some is still considered fatal. Here's an
    example:

            {
                    local $POE::Component::SSLify::IGNORE_SSL_ERRORS=1;
                    my $ctx = SSLify_CreateContext(...);
                    #Some more stuff
            }

 OpenSSL functions

    Theoretically you can do anything that Net::SSLeay exports from the
    OpenSSL libs on the socket. However, I have not tested every possible
    function against SSLify, so use them carefully!

  Net::SSLeay::renegotiate

    This function has been tested ( it's in t/2_renegotiate_client.t ) but
    it doesn't work on FreeBSD! I tracked it down to this security
    advisory:
    http://security.freebsd.org/advisories/FreeBSD-SA-09:15.ssl.asc which
    explains it in detail. The test will skip this function if it detects
    that you're on a broken system. However, if you have the updated
    OpenSSL library that fixes this you can use it.

    NOTE: Calling this means the callback function you passed in
    "Client_SSLify" or "Server_SSLify" will not fire! If you need this
    please let me know and we can come up with a way to make it work.

 Upgrading a non-ssl socket to SSL

    You can have a normal plaintext socket, and convert it to SSL anytime.
    Just keep in mind that the client and the server must agree to sslify
    at the same time, or they will be waiting on each other forever! See
    t/3_upgrade.t for an example of how this works.

 Downgrading a SSL socket to non-ssl

    As of now this is unsupported. If you need this feature please let us
    know and we'll work on it together!

 MSWin32 is not supported

    This module doesn't work on MSWin32 platforms at all ( XP, Vista, 7,
    etc ) because of some weird underlying fd issues. Since I'm not a
    windows developer, I'm unable to fix this. However, it seems like
    Cygwin on MSWin32 works just fine! Please help me fix this if you can,
    thanks!

 LOAD_SSL_ENGINES

    OpenSSL supports loading ENGINEs to accelerate the crypto algorithms.
    SSLify v1.004 automatically loaded the engines, but there was some
    problems on certain platforms that caused coredumps. A big shout-out to
    BinGOs and CPANTesters for catching this! It's now disabled in v1.007
    and you would need to explicitly enable it.

            sub POE::Component::SSLify::LOAD_SSL_ENGINES () { 1 }
            use POE::Component::SSLify qw( Client::SSLify );

EXPORT

    Stuffs all of the functions in @EXPORT_OK so you have to request them
    directly.

SEE ALSO

    Please see those modules/websites for more information related to this
    module.

      * POE

      * Net::SSLeay

SUPPORT

 Perldoc

    You can find documentation for this module with the perldoc command.

      perldoc POE::Component::SSLify

 Websites

    The following websites have more information about this module, and may
    be of help to you. As always, in addition to those websites please use
    your favorite search engine to discover more resources.

      * MetaCPAN

      A modern, open-source CPAN search engine, useful to view POD in HTML
      format.

      http://metacpan.org/release/POE-Component-SSLify

      * Search CPAN

      The default CPAN search engine, useful to view POD in HTML format.

      http://search.cpan.org/dist/POE-Component-SSLify

      * RT: CPAN's Bug Tracker

      The RT ( Request Tracker ) website is the default bug/issue tracking
      system for CPAN.

      http://rt.cpan.org/NoAuth/Bugs.html?Dist=POE-Component-SSLify

      * AnnoCPAN

      The AnnoCPAN is a website that allows community annotations of Perl
      module documentation.

      http://annocpan.org/dist/POE-Component-SSLify

      * CPAN Ratings

      The CPAN Ratings is a website that allows community ratings and
      reviews of Perl modules.

      http://cpanratings.perl.org/d/POE-Component-SSLify

      * CPAN Forum

      The CPAN Forum is a web forum for discussing Perl modules.

      http://cpanforum.com/dist/POE-Component-SSLify

      * CPANTS

      The CPANTS is a website that analyzes the Kwalitee ( code metrics )
      of a distribution.

      http://cpants.cpanauthors.org/dist/overview/POE-Component-SSLify

      * CPAN Testers

      The CPAN Testers is a network of smokers who run automated tests on
      uploaded CPAN distributions.

      http://www.cpantesters.org/distro/P/POE-Component-SSLify

      * CPAN Testers Matrix

      The CPAN Testers Matrix is a website that provides a visual overview
      of the test results for a distribution on various Perls/platforms.

      http://matrix.cpantesters.org/?dist=POE-Component-SSLify

      * CPAN Testers Dependencies

      The CPAN Testers Dependencies is a website that shows a chart of the
      test results of all dependencies for a distribution.

      http://deps.cpantesters.org/?module=POE::Component::SSLify

 Email

    You can email the author of this module at APOCAL at cpan.org asking
    for help with any problems you have.

 Internet Relay Chat

    You can get live help by using IRC ( Internet Relay Chat ). If you
    don't know what IRC is, please read this excellent guide:
    http://en.wikipedia.org/wiki/Internet_Relay_Chat. Please be courteous
    and patient when talking to us, as we might be busy or sleeping! You
    can join those networks/channels and get help:

      * irc.perl.org

      You can connect to the server at 'irc.perl.org' and join this
      channel: #perl-help then talk to this person for help: Apocalypse.

      * irc.freenode.net

      You can connect to the server at 'irc.freenode.net' and join this
      channel: #perl then talk to this person for help: Apocal.

      * irc.efnet.org

      You can connect to the server at 'irc.efnet.org' and join this
      channel: #perl then talk to this person for help: Ap0cal.

 Bugs / Feature Requests

    Please report any bugs or feature requests by email to
    bug-poe-component-sslify at rt.cpan.org, or through the web interface
    at http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE-Component-SSLify.
    You will be automatically notified of any progress on the request by
    the system.

 Source Code

    The code is open to the world, and available for you to hack on. Please
    feel free to browse it and play with it, or whatever. If you want to
    contribute patches, please send me a diff or prod me to pull from your
    repository :)

    https://github.com/apocalypse/perl-poe-sslify

      git clone https://github.com/apocalypse/perl-poe-sslify.git

AUTHOR

    Apocalypse <APOCAL@cpan.org>

ACKNOWLEDGEMENTS

            Original code is entirely Rocco Caputo ( Creator of POE ) -> I simply
            packaged up the code into something everyone could use and accepted the burden
            of maintaining it :)
    
            From the PoCo::Client::HTTP code =]
            # This code should probably become a POE::Kernel method,
            # seeing as it's rather baroque and potentially useful in a number
            # of places.

    ASCENT also helped a lot with the nonblocking mode, without his hard
    work this module would still be stuck in the stone age :)

    A lot of people helped add various features/functions - please look at
    the changelog for more detail.

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Apocalypse.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.

    The full text of the license can be found in the LICENSE file included
    with this distribution.

DISCLAIMER OF WARRANTY

    THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY
    APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT
    HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM "AS IS" WITHOUT
    WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT
    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
    PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE
    OF THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE DEFECTIVE, YOU
    ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR
    CONVEYS THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES,
    INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES
    ARISING OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT
    NOT LIMITED TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES
    SUSTAINED BY YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO
    OPERATE WITH ANY OTHER PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY
    HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.



Changes
Revision history for POE-Component-SSLify

1.011 2014-11-14T03:06:48Z UTC
 - Tweak the testsuite thanks to CPANTesters!

 [STATISTICS]
 - code churn: 11 files changed, 37 insertions(+), 35 deletions(-)

1.010 2014-11-12T06:13:41Z UTC
 - add informational test that prints out ssleay version
 - clamp the writes to 16K internally to get around write errors reported
   by RT#95071 and RT#58243, thanks!

 [STATISTICS]
 - code churn: 7 files changed, 591 insertions(+), 5 deletions(-)

1.009 2014-11-12T00:41:25Z UTC
 - Updates to the auto-generated files
 - Tweaked the testsuite to use done_testing() for sanity, RT#66741 (thanks
   RCAPUTO!)
 - Use Test::FailWarnings in the testsuite in place of Test::NoWarnings
   (DAGOLDEN++)
 - Add t/simple_large.t to try and track down RT#58243 (thanks Johan!)
 - Added a boatload of tests to try and track down RT#95071 (thanks Joe!)

 [STATISTICS]
 - code churn: 27 files changed, 2160 insertions(+), 1503 deletions(-)

1.008 2011-05-04T21:55:27Z UTC
 - Bazerka@irc noticed that POE wasn't even listed as a runtime prereq,
   d0h!

1.007 2011-05-04T21:36:32Z UTC
 - Revert the ENGINE load logic added in 1.004 because it was causing
   coredumps on netbsd ( thanks BINGOS! )
 - Add the constant LOAD_SSL_ENGINES sub if you need to load the ssl
   engines
 - Tweak the renegotiate tests to be more tolerant of failures, thanks
   RCAPUTO - RT#66741

1.006 2011-04-22T00:21:51Z UTC
 - Fix a heinous bug where openssl will randomly+reliably close a
   connection ( thanks MIRE )
 - Tweak the hook tests so they test the case where no data is sent over
   the connection
 - Tweak the connfail tests so they work better

1.005 2011-03-10T07:55:30Z UTC
 - Fix the connfail tests due to an incorrect assumption regards socket
   close value, thanks CPANTesters!

1.004 2011-03-08T23:39:03Z UTC
 - We now load certificate files via CTX_use_certificate_chain_file(),
   thanks Zephaniah E. Loss-Cutler-Hull <warp-spam_perl@aehallh.com>
   OpenSSL docs suggest it -
   http://www.openssl.org/docs/ssl/SSL_CTX_use_certificate.html#NOTES
   PLEASE yell at me if you need the old functionality - the docs suggest
   this is the "better" way to do it...
 - Add the ability to pass a subref to call on connection/negotiation
   success, thanks Zephaniah E. Loss-Cutler-Hull
   <warp-spam_perl@aehallh.com> NOTE: This will not work if you do
   renegotiation or any other zany SSL stuff!
 - Add the SSLify_GetStatus function to get the status of the connection
 - After staring at the Net::SSLeay/OpenSSL docs for a while I realized we
   were missing support for sslv23 version, added!
 - After some investigation, we now load all default ENGINEs for OpenSSL on
   startup, as it might provide a performance boost

1.003 2011-02-28T15:52:24Z UTC
 - Add $IGNORE_SSL_ERRORS to ignore certain SSL errors, thanks MNUNBERG -
   RT#66130
 - Remove prereq on vars.pm as it's obsolete

1.002 2011-02-20T04:23:48Z UTC
 - Add SSLify_GetSSL to get the Net::SSLeay object, thanks MNUNBERG!

1.001 2011-02-13T18:11:07Z UTC
 - Fixed the Net::SSLeay import code, thanks CPANTesters!

1.000 2011-02-13T00:06:14Z UTC
 - Converted to Dist::Zilla for the release process!
 - Add a test for in-situ sslification, thanks MNUNBERG!
 - Add prereq for IO::Handle 1.28 to get sane $socket->blocking( 0 )
   behavior on MSWin32
 - Remove crufty old code for nonblocking and use $socket->blocking()
   instead

0.20 0000-00-00T00:00:00Z UTC
 - Split up the simple.t test into 2 tests for clarity, and added more diag
   messages for renegotiate, thanks HMBRAND!

0.19 0000-00-00T00:00:00Z UTC
 - Fixed a warning generated by POE::Component::Client::TCP in t/simple.t,
   thanks HMBRAND!

0.18 0000-00-00T00:00:00Z UTC
 - Bumped POE dep to at least 1.267 for t/simple.t - thanks CPANTesters!
 - Minor typo fixes in POD/Build.PL

0.17 0000-00-00T00:00:00Z UTC
 - Fixed the t/simple.t test to PASS on FreeBSD because
   Net::SSLeay::renegotiate was buggy on it, thanks CPANTesters!
 - Added note about OpenSSL functions in the POD.

0.16 0000-00-00T00:00:00Z UTC
 - Updated the nonblocking code to be production-ready, thanks ASCENT!
 - Removed the NONBLOCKING() sub, this module is now always nonblocking.
 - Added more tests, thanks ASCENT!
 - Added "mylib/example.crt" and "mylib/example.key" for testing, thanks
   ASCENT!
 - Misc kwalitee and POD fixes.
 - Bumped Net::SSLeay prereq to 1.36 so we have the latest SSL stuff to
   ensure sanity :)

0.15 0000-00-00T00:00:00Z UTC
 - Added "examples/serverclient.pl" to track down same-process sslification
   problems, thanks LotR!
 - Applied patch from BinGOs to support passing custom $ctx for
   Server_SSLify, thanks! RT#43018
 - Switched over to Test::Apocalypse for easy author tests
 - Added experimental NONBLOCKING code, thanks ASCENT for the motivation!

0.14 0000-00-00T00:00:00Z UTC
 - removed Test::* modules from dependency list, thanks BINGOS - RT #36725
 - dos2unix fixes - thanks RT #36704
 - added Build.PL

0.13 0000-00-00T00:00:00Z UTC
 - POD typo errors in SSLify_ContextCreate - thanks ASCENT!

0.12 0000-00-00T00:00:00Z UTC
 - Kwalitee-related fixes

0.11 0000-00-00T00:00:00Z UTC
 - allowed setting of client-side context ( $ctx ) object - thanks RT
   #34442
 - squashed typo in pod - thanks ASCENT!
 - changed version check code to regexp for compatibility with SSLeay
   v1.33_01 - thanks Mark!
 - added SSLify_ContextCreate helper function
 - backported Net::SSLeay's removal of %Filenum_Objects hash

0.10 0000-00-00T00:00:00Z UTC
 - More tweaks of POD - finally close RT #31238
 - Added SSL version support - thanks RT #31492
 - Added SSL CTX option support as a side effect
 - Added client.pl example with ReadLine support

0.09 0000-00-00T00:00:00Z UTC
 - Minor tweak of POD to enable better distro building - thanks RT #31238

0.08 0000-00-00T00:00:00Z UTC
 - Added support for BINMODE - thanks RT #27117

0.07 0000-00-00T00:00:00Z UTC
 - Fixed undefined $info - thanks RT #22372

0.06 0000-00-00T00:00:00Z UTC

Makefile.PL
#
# This file is part of POE-Component-SSLify
#
# This software is copyright (c) 2014 by Apocalypse.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
use strict; use warnings;
# This Makefile.PL for POE-Component-SSLify was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.014
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.31.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'ExtUtils::MakeMaker' => '0',
    'Module::Build::Tiny' => '0.039',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.006;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Makes using SSL in the world of POE easy!",
  "AUTHOR" => "Apocalypse <APOCAL\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Module::Build::Tiny" => "0.039"
  },
  "DISTNAME" => "POE-Component-SSLify",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "POE::Component::SSLify",
  "PREREQ_PM" => {
    "Exporter" => 0,
    "IO::Handle" => "1.28",
    "Net::SSLeay" => "1.36",
    "POE" => "1.267",
    "Scalar::Util" => 0,
    "Symbol" => 0,
    "Task::Weaken" => "1.03",
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "IO::Handle" => "1.28",
    "IPC::Open3" => 0,
    "POE" => "1.267",
    "POE::Component::Client::TCP" => 0,
    "POE::Component::Server::TCP" => 0,
    "POE::Filter::Stream" => 0,
    "Socket" => 0,
    "Test::FailWarnings" => 0,
    "Test::More" => "1.001002"
  },
  "VERSION" => "1.011",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "IO::Handle" => "1.28",
  "IPC::Open3" => 0,
  "Module::Build::Tiny" => "0.039",
  "Net::SSLeay" => "1.36",
  "POE" => "1.267",
  "POE::Component::Client::TCP" => 0,
  "POE::Component::Server::TCP" => 0,
  "POE::Filter::Stream" => 0,
  "Scalar::Util" => 0,
  "Socket" => 0,
  "Symbol" => 0,
  "Task::Weaken" => "1.03",
  "Test::FailWarnings" => 0,
  "Test::More" => "1.001002",
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
